### ISSUE TITLE
Move `BigFloat` to stdlib

### ISSUE THREAD

**Issue by simonbyrne** [JuliaLang/julia#25716]
`BigFloat` could be moved to a stdlib package (`BigFloats`? `MPFR`?), though there are a few places it is coupled to Base:
- [ ] `widen(::Float64)` and `big(::AbstractFloat)`
  * Should we just leave these undefined, and pirate them in the package?
- [ ] array hashing ([which calls `widen`](https://github.com/JuliaLang/julia/blob/6aae4d5b3f664bb3ceafc71814007bc733a204f6/base/abstractarray.jl#L1992))
  * Should be doable via `TwicePrecision` tricks
- [ ] `@irrational` macro, which defines a [`BigFloat` method](https://github.com/JuliaLang/julia/blob/6aae4d5b3f664bb3ceafc71814007bc733a204f6/base/irrationals.jl#L130-L139)
- [ ] [`Float64(::Irrational,::RoundingMode)`](https://github.com/JuliaLang/julia/blob/6aae4d5b3f664bb3ceafc71814007bc733a204f6/base/irrationals.jl#L47-L51) (which is used in floating point inequalities)
- [ ] [`Rational(::Irrational)`](https://github.com/JuliaLang/julia/blob/6aae4d5b3f664bb3ceafc71814007bc733a204f6/base/irrationals.jl#L31-L45)

c.f. https://discourse.julialang.org/t/moving-bigfloat-to-stdlib/8545

**Comment by ViralBShah** [comment]
We should definitely move BigFloat to stdlib.

**Comment by simonbyrne** [comment]
One possible solution to the `Irrational` ones is that we manually define the necessary methods for the built-in ones (pi, e, etc.), and then move the `@irrational` macro itself to the new package?

**Comment by JeffBezanson** [comment]
Yes, that sounds ok. The type piracy here also does not bother me. Initially you have no way to widen a Float64, but if you load the BigFloats package then you do. Makes sense to me.

**Comment by nalimilan** [comment]
As I noted [on Discourse](https://discourse.julialang.org/t/moving-bigfloat-to-stdlib/8545/5?u=nalimilan), I'm afraid we can't define `widen(::Float64)` in the stdlib, or you wouldn't be able to hash heterogeneous arrays containing `Float64` values without loading the BigFloats stdlib module. The reasons for this are complex, and so far we haven't found any alternative (https://github.com/JuliaLang/julia/pull/16401/). However, we could move all user-facing APIs to the stdlib and only use `BigFloat` internally.

**Comment by StefanKarpinski** [comment]
I don't see a ton of advantage to this. We basically need to have BigInt in Base since we have syntax for it (long integer literals). We only export `BigFloat` itself and `setprecision` from the MPFR module, so it's not like it's a big contributor to Base API bloat. We already have the `big` function which we would like to have continue to work on floats without needing to load `BigFloats` so we need to have the BigFloat functionality loaded in the system image no matter what. So it seems like the only real change would be that if people want to change the default BigFloat precision they would have to write `using BigFloats` before doing `setprecision`. However, the `setprecision` function seems to aim to be generic – although it also only seems to currently support changing the precision of `BigFloat` and I'm not sure what else it could even change the precision of.

**Comment by pkofod** [comment]
> nd I'm not sure what else it could even change the precision of.

other arbitrary precision numbers? (I mean, packages might extend the function with a method for arbfloats, ...)

**Comment by StefanKarpinski** [comment]
Right, so we want a shared extension point anyway, which means to me that `setprecision` doesn't even really belong to `BigFloats` – it just happens to be defined there because `BigFloat` is the only Base type that it applies to. So really the *only* thing that would be exported from a `BigFloats` standard library would be the `BigFloat` type itself.

**Comment by simonbyrne** [comment]
The main advantage as I see it is that the MPFR dependency could eventually be made optional.

**Comment by JeffBezanson** [comment]
Agreed, I see this as a step towards making MPFR fully optional.

**Comment by ViralBShah** [comment]
Is this still something we can accomplish?

**Comment by ViralBShah** [comment]
A good first step is to remove from Base and make an stdlib. Can we do both BigFloat and BigInt as stdlibs? That paves the way for eventual separation.

**Comment by JeffreySarnoff** [comment]
BigInt and BigFloat share their first three letters, and that is their most alike quality.
stdlib `BigNums.jl` could export both.  Where one goes, in a tree-like sense, the other should follow.

**Comment by KristofferC** [comment]
The first step is to figure out how to deal with the fact that e.g. big integers are a part of the *syntax*  of the language:

```
julia> Meta.parse("10000000000000000000") |> dump
Expr
  head: Symbol macrocall
  args: Array{Any}((3,))
    1: GlobalRef
      mod: Module Core
      name: Symbol @int128_str
    2: Nothing nothing
    3: String "10000000000000000000"
```

What should this mean when BigInt is an stdlib and might not be available? Do you need to add a dependency to BigInts if you write big number literals in your code?

**Comment by JeffreySarnoff** [comment]
It seems asymmetric that 
```
abigint = 123456789_123456789_123456789_123456789_123456789
typeof(abigint) == BigInt

abigfloat = 123456789_123456789_123456789_123456789.123456789_12345678
typeof(abigfloat) == Float64
```

`BigInt("<digits>")` does not do anything, however
to write a BigFloat constant into a source file, the type and a string are needed
     - it is an oversight that this does not recognize underscores for clarity (`12345_67890`)
     - `BigFloat("<digits>[.<digits>][e±<digits>]")`
```
abigfloat = BigFloat("123456789123456789123456789123456789.987654321987654321")
typeof(abigfloat) == BigFloat
```

Writing  numeric constants with many digits in a source file is error-prone without support for underscores.
Writing BigInts one way and BigFloats another way to use them in a source file is kinda obtuse,

*there is much more efficient way to remove or to ignore the '_'s -- its not something I have had to do*
```
removeunderscore(s::AbstractString) = join(split(s, '_'))

BigInt(s::AbstractString) = parse(BigInt,  removeunderscore(s))
BigFloat(s::AbstractString) = parse(BigFloat, removeunderscore(s))
```

**Comment by ViralBShah** [comment]
@JeffreySarnoff Can you file that as a separate issue?

**Comment by oscardssmith** [comment]
This isn't fully asymmetric because almost all decimal numbers require an infinite number of bits to represent exactly in binary.

**Comment by JeffreySarnoff** [comment]
https://github.com/JuliaLang/julia/issues/45760

Oscar, I am focused on the uncoupled (dissonant) nature of their approaches to recognizing "typed" constants.

**Comment by ViralBShah** [comment]
So maybe close this issue then?

**Comment by KristofferC** [comment]
I was only talking about BigInt here. BigFloat should be movable I think.

**Comment by ViralBShah** [comment]
Wouldn't it be a bit odd though to have one as a stdlib and not the other, since then you have to know which one needs to be declared as a depencency and which one is not?

**Comment by KristofferC** [comment]
One difference is that you can get a dependency on BigInt by just writing a big integer literal but to get a dependency on BigFloat you need to use something explicit e.g. the `big""` string macro.

**Comment by PallHaraldsson** [comment]
>I don't see a ton of advantage to this. We basically need to have BigInt in Base since we have syntax for it (long integer literals). We only export BigFloat itself and setprecision from the MPFR module, so it's not like it's a big contributor to Base API bloat. We already have the big function which we would like to have continue to work on floats

I don't want `big` to work on floats. Before we discuss BigInt which is a separate issue, I would simply like `BigFloat`s gone to an external MPFR.jl library but without it being a stdlib.

What are the pros and cons of it as a stdlib, or included at all? I want to suggest users use ArbNumerics.jl OR MPFR.jl but not usually at the same time so I could do:

const BigFloat = ArbNumerics.ArbFloat

i.e. at that to that package, and it would work as a faster drop-in replacement, for most users.

**Comment by PallHaraldsson** [comment]
>widen(::Float64) and big(::AbstractFloat)
>Should we just leave these undefined, and type-pirate them in the package?

undefined, or I guess you rather mean:

widen(Float64) = error("Install ArbNumerics.jl (or MPFR.jl) for this to work)"

@simonbyrne
>array hashing (which calls widen)

couldn't that, and all other cases you mention be handled by that erroring definition and/or defining in a package, pirating?

E.g.
>@irrational macro, which defines a BigFloat method

it's not clear it's best to get BigFloat out (with its default 256 bits), the next step for widen should be to 128 bits I recall you get for ArbNumerics.ArbFloat, its default is settable, to e.g. 256 bits, hypothetically you could widen to higher and higher or to MPFR.

The first step is to figure out how to deal with the fact that e.g. big integers are a part of the syntax of the language:

julia> Meta.parse("10000000000000000000") |> dump
[..]
      name: Symbol @int128_str

It seems you made an invalid point since you get Int128 up to:

Meta.parse("170141183460469231731687303715884105728")  # On higher than the typemax, there's also UIn128.

I think we could just error in the parser and it wouldn't be awfully bad. And support for BigInt can be dynamically added to the parser(?), Julia doesn't need BigInt for itself (in the parser). It's possibly easier to add to the parser at runtime with JuliaSyntax.jl now used, if that's actually advised...

@JeffreySarnoff
>BigInt and BigFloat share their first three letters, and that is their most alike quality.
stdlib BigNums.jl could export both.

The reason I don't like this much is that the users that want BigInt may not need or want BigFloat and vice versa. Eventually we should have a BigInt (external or stdlib) library (or name it GMP.jl?), in addition to for `BigFloat`s. It only seems to make sense to add a BigNums.jl (stdlib) to include both, in case you want to actually use both, and for for sure nor ArbNumerics.jl inseaed of MPFR.jl for the BigFloat implementation.

We could have `using Julia1` to restore full compatibility.

**Comment by nalimilan** [comment]
FWIW array hashing no longer uses `widen` since https://github.com/JuliaLang/julia/pull/26022.
